### ISSUE TITLE
Request to use constants for creating/dispatching/handling actions

### ISSUE THREAD

**Issue by mbrio** [malte-wessel/minimal-flux#6]
I noticed the other day you had starred my Tweet for React Boilerplate and I wanted to thank you. It also got me to thinking that out of all the Flux libraries I have been playing with I wanted to like yours the best. I thought the API was easy to understand, clean, and mostly explicit but I found the way in which you are handling action names to be muddying up minimal-flux's simplicity.

I found it to be a bit questionable that your dispatcher was transforming your action's event name by concatenating it with a camel-cased version of the emitters constructor. I feel as though taking a tip from MartyJS and creating a `Constant` class would be the best solution as it would be much more explicit and easier to understand at first glance.

I've forked minimal-flux and did just that: https://github.com/mbrio/minimal-flux

I didn't want to send you a PR because the API is the only thing complete. I didn't add tests or documentation because if this was something you weren't interested in I didn't want to waste time.

In essence my fork allows for: `this.dispatch('create', message);` to be changed to `this.dispatch(messageConstants.create, message);`; and `this.handleAction('messages.create', this.handleCreate);` to be changed to `this.handleAction(messageConstants.create, this.handleCreate);`. I feel this is much clearer as actions and stores would utilize the same **exact** key to describe an action and it's handler.

I did something I did not like to do in order to get this to work and that is to make your `Actions` class `emit` method only emit a single event (*fluxAction*). This event passes the constant representing the action as the first parameter to the event handler, and any other parameters there after.

I think a better solution would be to make a breaking change that would force users to use only the `dispatch` function (as opposed to also allowing the `emit` function), and have the dispatch function emit this event.

If this is something you'd like to move forward with I'll be happy to add tests and documentation.

**Comment by malte-wessel** [comment]
Thanks for your feedback! I really like your boilerplate and the way you integrated minimal-flux!
You're right, it's sometimes more clearer to have one and the same exact key for both the dispatch and the handleAction method. I had a similar implementation in a previous version but without the constant class. For each action the dispatcher created an actionId that you had to pass to the handleAction method like:

````javascript
// Inside a store
let actionId = this.actionIds.messages.create
this.handleAction(actionId, handler);
`````

However, I felt like there was some overhead with accessing the actionId and pass it to the handleAction method, so I decided to stick with plain strings. 

I also [discussed some patterns here](https://github.com/malte-wessel/minimal-flux/issues/1#issuecomment-78464177) to make the dispatch method easier to use.

I'm not sure how I feel about Constants. When I was playing with the original flux library I always thought that Constants introduce too much bookkeeping [while the benefits are quite small](https://groups.google.com/d/msg/reactjs/dzVhoavALwI/eeezh-kKsr0J).

Regarding your implementation: how would the constants be accessed (e.g. `this.constants.message.create`) and how would they be created? By the dispatcher or by hand?

**Comment by mbrio** [comment]
In my implementation there is a minimal amount of magic in order to produce a more explicit interface. The developer creates constants with `export default Constant.createConstants('action1', 'action2')`. This returns an object containing two properties (`action1` and `action2`) that point to `Constant` instances. In any source needing access to the constants you would `import appConstants from './AppConstants';` and use them like `appConstants.action1`. The only magic is the magic you've put in place that causes there to be a function on the `Actions` class that corresponds with the action name.

I was thinking that in addition to this your dispatcher may gain a property called `constants` that you can set when creating the apps flux class, you would set it at the time you set stores and actions.

```
import { Flux } from 'minimal-flux';
import SomeActions from './actions/SomeActions';
import SomeStore from './actions/SomeStore';
import someConstants from './actions/SomeConstants';

export default new Flux({
  actions: {
    someActions: SomeActions
  },
  stores: {
    someStore: SomeStore
  },
  constants: {
    someConstants: someConstants
  }
});
```

And sorry if I made any mistakes, I'm unfortunately writing this on the train with my phone.

**Comment by mbrio** [comment]
As a follow up I've refactored some of my code at https://github.com/mbrio/minimal-flux, I've added comments and documentation, as well as made all tests passing. Taking some code from *test/Dispatcher.js*:

```
let fooConstants = Constant.createConstants('foo', 'bar', 'baz');

class FooStore extends Store {

    constructor() {
        this.handleAction(fooConstants.foo, function() {
            flux.actions.foo.bar();
        });
    }

}

class FooActions extends Actions {
    foo() {
        this.dispatch(fooConstants.foo);
    }
    bar() {
        this.dispatch(fooConstants.bar);
    }
}
```

**Comment by malte-wessel** [comment]
I tend to not introducing another class. However, what do you think about automatically creating constants from the passed actions. You could the use them like this:

````javascript
class MessagesActions extends Actions {
  create(msg) {
    this.dispatch(this.constants.messages.create, msg)
  }
}

class MessagesStore extends Store {
  constructor() {
    this.handleAction(this.constants.messages.create, this.handleCreate);
  }
}
````

**Comment by mbrio** [comment]
Hi Malte,

That's better than what you currently have, is there a reason why you are adding `messages` to `constants`. I think that auto-magic naming based off of the constructor is what makes it confusing. If it could just be `this.constants.create` or even more explicitly as `this.constants.MessagesActions.create`. I was playing around with a structure that just took `MessagesActions.create`, similar to how Reflux handles it. I liked it, it felt very explicit without adding another class.

**Comment by malte-wessel** [comment]
The idea behind automatically generating the constants is that you don't have to manage another set of classes. The generating will be based on the names of the actions. You access your actions already via `flux.actions.messages.create`. The constants will have the same path e.g. `flux.constants.messages.create`. IMHO that's the easiest way to remember your entities. I'm currently working on an implementation. I will publish a branch for it soon. So long!

**Comment by mbrio** [comment]
I wasn't suggesting that the automatic creation of constants was a bad idea, I think that's good; I guess I'm only worried about the naming of the property on the constants property. So am I to assume with this setup you'd configure your flux like this:

```
let flux = new Flux({
  actions: {
    messages: AnActionsClass
  },
  stores: {
    messages: AStoreClass
  }
});
```

And you'd end up with access to `flux.constants.messages` which maps to the methods of the actions class? Does that mean that the constants property name would be whatever the actions property name is. For example:

```
let flux = new Flux({
  actions: {
    messages: AnActionsClass
  },
  stores: {
    somethingOtherThanMessages: AStoreClass
  }
});
```

would give access to `flux.constants.messages` still?

If that's the case it sounds good because the naming of the constant isn't magical because it's defined by the user via the creation of the actions property.

**Comment by malte-wessel** [comment]
Yes, exactly. The constants will be created from the actions namespace e.g. `messages` and it's methods.

````javascript

class MessagesActions extends Actions {
  create() {
    this.dispatch(this.constants.messages.create);
  }
}

class AwesomeStore extends Store {
  constructor() {
    this.handleAction(this.constants.messages.create, this.handleMessagesCreate);
  }
  // ...
}

let flux = new Flux({
  actions: {
    messages: MessagesActions
  },
  stores: {
    awesome: AwesomeStore
  }
});
````

**Comment by mbrio** [comment]
Great, this is much clearer at first glance than the current implementation. My only issue (which is minor and I understand why you don't want constants) is that by having an auto-generated constant object namespaced by how the actions are configured in the Flux instance, you are tightly coupling your stores to a specific Flux naming scheme. It's very minor, and I think the cleanliness of this structure is worth the tradeoff.
